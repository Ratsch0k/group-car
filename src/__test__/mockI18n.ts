jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () => new Promise(() => undefined),
        languages: ['de', 'en'],
        language: 'en',
      },
    };
  },
}));
