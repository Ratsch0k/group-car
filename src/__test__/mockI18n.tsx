import React from 'react';

jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () => new Promise(() => undefined),
        languages: ['de', 'en'],
        language: 'en',
      },
    };
  },
  Trans: (props: any) => {
    return <div>{props.i18nKey}</div>;
  },
}));
