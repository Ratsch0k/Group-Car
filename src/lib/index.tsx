export * from './components';
export * from './ModalRouter';
export * from './theme';
export * from './requests';
export * from './Routes';
export * from './context';
export * from './hooks';
