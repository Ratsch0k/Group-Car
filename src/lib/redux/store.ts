import {configureStore, getDefaultMiddleware} from '@reduxjs/toolkit';
import rootReducers from './slices';
import {
  routerMiddleware,
} from 'connected-react-router';
import history from 'lib/redux/history';
import * as Sentry from '@sentry/react';

const sentryReduxEnhancer = Sentry.createReduxEnhancer({
  actionTransformer: (action) => {
    return action;
  },
  stateTransformer: (state) => {
    return state;
  },
});


/**
 * Create redux store.
 */
const store = configureStore({
  reducer: rootReducers(history),
  devTools: true,
  middleware: getDefaultMiddleware().concat(routerMiddleware(history)),
  enhancers: [sentryReduxEnhancer],
});

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;

export default store;
