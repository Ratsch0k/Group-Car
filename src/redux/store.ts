import {configureStore} from '@reduxjs/toolkit';
import rootReducer from './slices';

/**
 * Create redux store.
 */
const store = configureStore({
  reducer: rootReducer,
  devTools: true,
});

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;

export default store;
