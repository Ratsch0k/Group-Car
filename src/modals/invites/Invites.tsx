import {
  Box,
  Dialog,
  DialogContent,
  Divider,
  List,
  Theme,
  Typography,
} from '@material-ui/core';
import {createStyles, makeStyles} from '@material-ui/styles';
import {CloseableDialogTitle, useModalRouter} from 'lib';
import {useInvites} from 'lib/hooks/useInvites';
import React, {useEffect} from 'react';
import {useTranslation} from 'react-i18next';
import InvitesListItem from './InvitesListItem';

/**
 * Styles.
 */
const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    emptyContainer: {
      display: 'grid',
      placeItems: 'center',
      padding: theme.spacing(4),
    },
  }),
);

export const Invites: React.FC = () => {
  const {invites, refresh, deleteInvite, acceptInvite} = useInvites();
  const {t} = useTranslation();
  const {close} = useModalRouter();
  const classes = useStyles();

  // Refresh invites on first render
  useEffect(() => {
    refresh();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  let content: JSX.Element;
  if (invites.length > 0) {
    content = (
      <List>
        {
          invites.map((invite, index) =>
            <React.Fragment key={`invite-container-${index}`}>
              {
                index > 0 &&
                <Divider />
              }
              <InvitesListItem
                invite={invite}
                delete={() => deleteInvite(invite.groupId)}
                accept={() => acceptInvite(invite.groupId)}
              />
            </React.Fragment>,
          )
        }
      </List>
    );
  } else {
    content = (
      <Box className={classes.emptyContainer}>
        <Typography color='textSecondary'>
          {t('modals.invites.empty')}
        </Typography>
      </Box>
    );
  }

  return (
    <Dialog open={true} fullWidth maxWidth='sm' onBackdropClick={close}>
      <CloseableDialogTitle close={close}>
        {t('modals.invites.title')}
      </CloseableDialogTitle>
      <DialogContent>
        {content}
      </DialogContent>
    </Dialog>
  );
};

export default Invites;
