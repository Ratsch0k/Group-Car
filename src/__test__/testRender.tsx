import React from 'react';
import {render, RenderResult} from '@testing-library/react';
import {ConnectedRouter} from 'connected-react-router';
import {Provider} from 'react-redux';
import history from '../lib/redux/history';
import {StylesProvider, ThemeProvider} from '@material-ui/styles';
import testTheme from './testTheme';
import configureMockStore, {
  MockStoreCreator,
  MockStoreEnhanced,
} from 'redux-mock-store';
import thunk from 'redux-thunk';
import {RootState} from '../lib/redux/store';
import {
  initialState as authState,
} from '../lib/redux/slices/auth/authSlice';
import {
  initialState as groupState,
} from '../lib/redux/slices/group/groupSlice';
import {
  initialState as invitesState,
} from '../lib/redux/slices/invites/invitesSlice';


export interface TestRenderResult<T> {
  mockStore: MockStoreCreator<T>;
  store: MockStoreEnhanced<unknown>;
}

const defaultState: RootState = {
  router: {
    action: 'POP',
    location: {
      pathname: '/',
      search: '',
      hash: '',
      query: {},
      state: undefined,
    },
  },
  auth: authState,
  group: groupState,
  invites: invitesState,
};

/**
 * Testing generateClassName function to
 * avoid weird fails due to random class names
 * @param rule Rule
 * @param stylesheet Stylesheet
 * @returns The class name
 */
const generateClassName = (rule, stylesheet) => {
  return `${stylesheet.options.classNamePrefix}-${rule.key}`;
};


/**
 * Render function for testing purposes.
 * Provides the necessary base components and
 * contexts.
 * @param children The children to render
 * @returns The result of the render function.
 */
function testRender<T>(state: T, children: React.ReactNode):
RenderResult & TestRenderResult<T> {
  const middleware = [thunk];
  const mockStore = configureMockStore<T>(middleware);
  const initialState = {
    ...defaultState,
    ...state,
  };
  const store = mockStore(initialState);

  const renderResult = render(
    <Provider store={store}>
      <ConnectedRouter history={history}>
        <ThemeProvider theme={testTheme}>
          <StylesProvider generateClassName={generateClassName}>
            {children}
          </StylesProvider>
        </ThemeProvider>
      </ConnectedRouter>
    </Provider>,
  );

  return {
    ...renderResult,
    mockStore,
    store,
  };
}

export default testRender;
