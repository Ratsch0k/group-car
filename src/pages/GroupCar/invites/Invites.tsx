import {
  Box,
  DialogContent,
  Divider,
  List, ModalProps,
  Theme,
  Typography,
} from '@material-ui/core';
import {createStyles, makeStyles} from '@material-ui/styles';
import {CloseableDialogTitle, Dialog} from 'lib';
import React, {useCallback, useEffect} from 'react';
import {useTranslation} from 'react-i18next';
import {useAppDispatch, useShallowAppSelector} from 'lib/redux/hooks';
import {closeModal} from 'lib/redux/slices/modalRouter/modalRouterSlice';
import InvitesListItem from './InvitesListItem';
import {
  getAllInvites,
  getInvites,
} from 'lib/redux/slices/invites';

/**
 * Styles.
 */
const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    emptyContainer: {
      display: 'grid',
      placeItems: 'center',
      padding: theme.spacing(4),
    },
  }),
);

/**
 * Invites modal for managing invites of user.
 */
export const Invites: React.FC = () => {
  const invites = useShallowAppSelector(getAllInvites);
  const {t} = useTranslation();
  const classes = useStyles();
  const dispatch = useAppDispatch();

  // Refresh invites on first render
  useEffect(() => {
    dispatch(getInvites());
    // eslint-disable-next-line
  }, []);

  const handleOnClose: ModalProps['onClose'] = useCallback((_event, reason) => {
    if (reason === 'backdropClick') {
      dispatch(closeModal());
    }
  }, []);

  let content: JSX.Element;
  if (invites.length > 0) {
    content = (
      <List>
        {
          invites.map((invite, index) =>
            <React.Fragment key={`invite-container-${index}`}>
              {
                index > 0 &&
                <Divider />
              }
              <InvitesListItem
                invite={invite}
              />
            </React.Fragment>,
          )
        }
      </List>
    );
  } else {
    content = (
      <Box className={classes.emptyContainer}>
        <Typography color='textSecondary' align='center'>
          {t('modals.invites.empty')}
        </Typography>
      </Box>
    );
  }

  return (
    <Dialog
      open={true}
      fullWidth
      maxWidth='sm'
      onClose={handleOnClose}
    >
      <CloseableDialogTitle close={() => dispatch(closeModal())}>
        {t('modals.invites.title')}
      </CloseableDialogTitle>
      <DialogContent>
        {content}
      </DialogContent>
    </Dialog>
  );
};

export default Invites;
