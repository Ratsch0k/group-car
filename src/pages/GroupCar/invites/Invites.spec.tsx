import mockedAxios from '../../../__test__/mockAxios';
import '../../../__test__/mockI18n';
import { InviteWithGroupAndInviteSender } from "../../../lib";
import {fireEvent, screen, waitFor} from '@testing-library/react';
import React from "react";
import Invites from './Invites';
import testRender from "../../../__test__/testRender";
import { RootState } from "../../../lib/redux/store";

describe('Invites modal', () => {
  let state: Partial<RootState>;
  let invites: InviteWithGroupAndInviteSender[];

  beforeEach(() => {
    invites = [
      {
        groupId: 1,
        userId: 1,
        Group: {
          id: 1,
          name: 'group-1',
          description: 'TEST',
          createdAt: new Date(),
          updatedAt: new Date(),
          ownerId: 10,
          Owner: {
            username: 'OWNER',
            id: 10,
          },
        },
        invitedBy: 10,
        InviteSender: {
          username: 'invite-sender-1',
          id: 10,
        },
        createdAt: new Date(),
        updatedAt: new Date(),
      },
      {
        groupId: 2,
        userId: 1,
        Group: {
          id: 2,
          name: 'group-2',
          description: 'TEST',
          createdAt: new Date(),
          updatedAt: new Date(),
          ownerId: 10,
          Owner: {
            username: 'OWNER',
            id: 10,
          },
        },
        invitedBy: 10,
        InviteSender: {
          username: 'invite-sender-2',
          id: 10,

        },
        createdAt: new Date(),
        updatedAt: new Date(),
      },
    ];

    state = {
      invites: {
        ids: invites.map((i) => i.groupId),
        entities: invites.reduce((l, n) => ({...l, [n.groupId]: n}), {}),
        loading: false,
      },
      auth: {
        user: {
          username: 'user',
          id: 1,
        },
        signUpRequestSent: false,
        loading: false,
      },
      group: {
        ids: [],
        entities: [],
        selectedGroup: null,
        loading: false,
      },
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders list of invites correctly', () => {
    const {baseElement} = testRender(
      state,
      <Invites />,
    );

    expect(screen.queryAllByText('misc.invitedBy')).toHaveLength(invites.length);
    expect(baseElement).toMatchSnapshot();
  });

  it('accept button dispatches acceptInvite action', async () => {
    mockedAxios.post = jest.fn().mockResolvedValue({data: undefined});
    mockedAxios.get = jest.fn().mockResolvedValue({data: {groups: []}});

    const {baseElement, store} = testRender(
      state,
      <Invites />,
    );

    expect(screen.queryAllByText('misc.invitedBy')).toHaveLength(invites.length);

    fireEvent.click(baseElement.querySelector(`#invite-${invites[0].groupId}-accept-btn`));

    const expectedPendingAction = {
      type: 'invites/acceptInvite/pending',
      payload: undefined,
      meta: {
        arg: invites[0].groupId,
        requestStatus: expect.any(String),
        requestId: expect.any(String),
      },
    };
    const expectedFulfilledAction = {
      type: 'invites/acceptInvite/fulfilled',
      payload: undefined,
      meta: {
        arg: invites[0].groupId,
        requestStatus: expect.any(String),
        requestId: expect.any(String),
      },
    };

    await waitFor(() => expect(store.getActions()).toContainEqual(expectedFulfilledAction));
    expect(store.getActions()).toContainEqual(expectedPendingAction);
    await waitFor(() => expect(mockedAxios.post).toHaveBeenCalledTimes(1));
    expect(mockedAxios.post).toHaveBeenCalledWith(`/api/user/invite/${invites[0].groupId}/join`);
  });
});
