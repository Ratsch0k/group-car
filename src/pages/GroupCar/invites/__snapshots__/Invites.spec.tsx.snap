// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Invites modal renders list of invites correctly 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper"
      role="none presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        class="MuiPaper-root MuiDialog-paper makeStyles-rounded MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth MuiPaper-elevation24 MuiPaper-rounded"
        role="dialog"
      >
        <div
          class="MuiDialogTitle-root makeStyles-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-wrap-xs-nowrap"
            >
              <div
                class="MuiGrid-root makeStyles-title MuiGrid-item"
              >
                modals.invites.title
              </div>
              <div
                class="MuiGrid-root makeStyles-closeButton MuiGrid-item"
              >
                <button
                  class="MuiButtonBase-root MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                      />
                    </svg>
                  </span>
                </button>
              </div>
            </div>
          </h2>
        </div>
        <div
          class="MuiDialogContent-root"
          style="min-height: 100px;"
        >
          <ul
            class="MuiList-root MuiList-padding"
          >
            <div
              class="MuiCollapse-root MuiCollapse-entered"
              style="min-height: 0px;"
            >
              <div
                class="MuiCollapse-wrapper"
              >
                <div
                  class="MuiCollapse-wrapperInner"
                >
                  <li
                    class="MuiListItem-container"
                  >
                    <div
                      class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
                    >
                      <div
                        class="MuiListItemText-root MuiListItemText-multiline"
                      >
                        <span
                          class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-colorTextPrimary MuiTypography-displayBlock"
                        >
                          group-1
                        </span>
                        <p
                          class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                        >
                          misc.invitedBy
                        </p>
                      </div>
                    </div>
                    <div
                      class="MuiListItemSecondaryAction-root"
                    >
                      <div
                        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1"
                      >
                        <div
                          class="MuiGrid-root MuiGrid-item"
                        >
                          <button
                            class="MuiButtonBase-root MuiIconButton-root makeStyles-button makeStyles-button"
                            id="invite-1-accept-btn"
                            tabindex="0"
                            title="misc.accept"
                            type="button"
                          >
                            <span
                              class="MuiIconButton-label"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
                                />
                              </svg>
                            </span>
                          </button>
                        </div>
                        <div
                          class="MuiGrid-root MuiGrid-item"
                        >
                          <button
                            class="MuiButtonBase-root MuiIconButton-root makeStyles-button makeStyles-button"
                            id="invite-1-delete-btn"
                            tabindex="0"
                            title="misc.delete"
                            type="button"
                          >
                            <span
                              class="MuiIconButton-label"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                                />
                              </svg>
                            </span>
                          </button>
                        </div>
                      </div>
                    </div>
                  </li>
                </div>
              </div>
            </div>
            <div
              class="MuiCollapse-root MuiCollapse-entered"
              style="min-height: 0px;"
            >
              <div
                class="MuiCollapse-wrapper"
              >
                <div
                  class="MuiCollapse-wrapperInner"
                >
                  <hr
                    class="MuiDivider-root"
                  />
                  <li
                    class="MuiListItem-container"
                  >
                    <div
                      class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
                    >
                      <div
                        class="MuiListItemText-root MuiListItemText-multiline"
                      >
                        <span
                          class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-colorTextPrimary MuiTypography-displayBlock"
                        >
                          group-2
                        </span>
                        <p
                          class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                        >
                          misc.invitedBy
                        </p>
                      </div>
                    </div>
                    <div
                      class="MuiListItemSecondaryAction-root"
                    >
                      <div
                        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1"
                      >
                        <div
                          class="MuiGrid-root MuiGrid-item"
                        >
                          <button
                            class="MuiButtonBase-root MuiIconButton-root makeStyles-button makeStyles-button"
                            id="invite-2-accept-btn"
                            tabindex="0"
                            title="misc.accept"
                            type="button"
                          >
                            <span
                              class="MuiIconButton-label"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
                                />
                              </svg>
                            </span>
                          </button>
                        </div>
                        <div
                          class="MuiGrid-root MuiGrid-item"
                        >
                          <button
                            class="MuiButtonBase-root MuiIconButton-root makeStyles-button makeStyles-button"
                            id="invite-2-delete-btn"
                            tabindex="0"
                            title="misc.delete"
                            type="button"
                          >
                            <span
                              class="MuiIconButton-label"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                                />
                              </svg>
                            </span>
                          </button>
                        </div>
                      </div>
                    </div>
                  </li>
                </div>
              </div>
            </div>
          </ul>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
